### DISCUSSION TITLE
[Feature] 关于接入第三方审批流

### DISCUSSION THREAD

**Discussion by boomboomxx** [hhyo/Archery#2757] [General]
你好,请问有接入第三方审批流的计划吗? 
类似**钉钉**, **飞书**, **企微** 这类都有常用的OA审批

或者有相关的改造流程说明或建议吗?

**Comment by LeoQuote** [comment]
欢迎pr

**Reply by boomboomxx** [reply]
目前我没有接入过相关的东西
我想确认一下如果需要改造的话, 具体的影响范围, 还有一个设计的规范和流程

**Reply by LeoQuote** [reply]
没有太多可以分享的, 可以直接看代码试试, 这方面的文档也比较少. 你可能要做的是一个适配器, 将飞书或其他 OA 系统的人员信息和 archery 的人员信息关联起来, 然后再做一个双向或单向的数据同步功能. 

首先当前有 generic webhook 功能, 可以将审批流, 执行等事件推送至外部, 然后OA 的审批流我猜测也是有各种外接的 webhook 功能, 或者轮询. archery 的 generic webhook 参见: https://github.com/hhyo/Archery/blob/bd229c2faaba27c115c54b5f8037e6087e5722d3/sql/notify.py#L99-L128

然后当前是有自定义审批流的功能: https://github.com/hhyo/Archery/blob/bd229c2faaba27c115c54b5f8037e6087e5722d3/sql/utils/workflow_audit.py#L128 你可以继承这个class 实现自己的 audit, 并在 https://github.com/hhyo/Archery/blob/bd229c2faaba27c115c54b5f8037e6087e5722d3/archery/settings.py#L71 这里进行配置, 即可做到自定义审批流.


这是当前有的一些工具和功能, 欢迎 pr, 只要对于不启用该功能的用户来说, 体验没有变化, 单元测试覆盖率达标, 风格上没有大问题, 一般都可以接受.

不过这个功能可能比较大, 我个人比较推荐做成一个独立外挂程序, 使用自定义审批流和自定义提醒的框架来做定制开发, 这个独立外挂程序可能还可以对接你们企业的其他执行系统.

**Reply by boomboomxx** [reply]
感谢建议!

之前看代码的时候看的是`Audit`, 有 `AuditV2` 之后需要修改的点就明确了.

目前我了解到不同平台基本都是有事件机制的, 接收不同平台的事件做工单处理就可以

我先看下实现一个的方案, 可能还是会基于已有的系统做审批集成, archery 用openapi调用实现
